form2request/_base.py

```diff
@@ -0,0 +1,220 @@
+            method = click_element.get("formmethod")
+        if method:
+            method_src = click_element
+        else:
+            method = form.method
+            assert method is not None  # lxml’s form.method is always filled
+            method_src = form
+    method = method.upper()
+    if method not in {"GET", "POST"}:
+        if method_src is USER:
+            raise ValueError(
+                f"The specified form method ({method!r}) is not supported."
+            )
+        if method_src is click_element:
+            raise NotImplementedError(
+                f"Found unsupported form method {method!r} in the formmethod "
+                f"attribute of the submission button."
+            )
+        raise NotImplementedError(f"Found unsupported form method {method!r}.")
+    return method
+
+
+class _NoClickables(ValueError):
+    pass
+
+
+def _click_element(
+    form: FormElement, click: None | bool | HtmlElement
+) -> HtmlElement | None:
+    if click is False:
+        return None
+    if click in {None, True}:
+        clickables = list(
+            form.xpath(
+                'descendant::input[re:test(@type, "^(submit|image)$", "i")]'
+                '|descendant::button[not(@type) or re:test(@type, "^submit$", "i")]',
+                namespaces={"re": "http://exslt.org/regular-expressions"},
+            )
+        )
+        if not clickables:
+            if click:
+                raise _NoClickables
+            else:
+                return None
+        click = clickables[0]
+    return click
+
+
+def _data(data: FormdataType, click_element: HtmlElement | None) -> FormdataType:
+    data = data or {}
+    if click_element is not None and (name := click_element.get("name")):
+        click_data = (name, click_element.get("value"))
+        if isinstance(data, dict):
+            data = dict(data)
+            data[click_data[0]] = click_data[1]
+        else:
+            data = list(data)
+            data.append(click_data)
+    return data
+
+
+def _query(form: FormElement, data: FormdataType) -> str:
+    keys = dict(data or ()).keys()
+    if not data:
+        data = []
+    inputs = form.xpath(
+        "descendant::textarea"
+        "|descendant::select"
+        "|descendant::input[not(@type) or @type["
+        ' not(re:test(., "^(?:submit|image|reset)$", "i"))'
+        " and (../@checked or"
+        '  not(re:test(., "^(?:checkbox|radio)$", "i")))]]',
+        namespaces={"re": "http://exslt.org/regular-expressions"},
+    )
+    values: list[FormdataKVType] = [
+        (k, "" if v is None else v)
+        for k, v in ((e.name, e.value) for e in inputs)
+        if k and k not in keys
+    ]
+    items = data.items() if isinstance(data, dict) else data
+    values.extend((k, v) for k, v in items if v is not None)
+    encoded_values = [
+        (k.encode(), v.encode())
+        for k, vs in values
+        for v in (cast("Iterable[str]", vs) if _is_listlike(vs) else [cast("str", vs)])
```

Comment:
Can it be handled on function definition level, via [typing.overload](https://docs.python.org/3/library/typing.html#overload)? Or is too much?

Reply:
I’ll try.

Reply:
Removes `_is_listlike` altogether.